File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        int n = arr.length;
        int[] pse = new int[n];
        int[] nse = new int[n];
        Stack<Integer> stack = new Stack<>();
        
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && arr[i] < arr[stack.peek()]) {
                stack.pop();
            }
            pse[i] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(i);
        }
        
        stack.clear();
        
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && arr[i] <= arr[stack.peek()]) {
                stack.pop();
            }
            nse[i] = stack.isEmpty() ? n : stack.peek();
            stack.push(i);
        }
        
        int mod = 1000000007;
        long sumSubarrayMins = 0;  
        
        for (int i = 0; i < n; i++) {
            long sum = (long) arr[i] * (i - pse[i]) * (nse[i] - i) ;
            sumSubarrayMins = (sumSubarrayMins + sum) % mod;
        }
        
        return (int) sumSubarrayMins;
    }
}

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        n = len(arr)
        MOD = 10**9 + 7
        
        # left[i] = number of consecutive elements to the left of (and including) arr[i] that are greater than or equal to arr[i]
        left = [0] * n
        stack = []
        for i in range(n):
            while stack and arr[i] < arr[stack[-1]]:
                stack.pop()
            
            if not stack:
                left[i] = i + 1
            else:
                left[i] = i - stack[-1]

            stack.append(i)

        # right[i] = number of consecutive elements to the right of (and including) arr[i] that are greater than arr[i]
        right = [0] * n
        stack = []
        for i in range(n - 1, -1, -1):
            while stack and arr[i] <= arr[stack[-1]]:
                stack.pop()
            
            if not stack:
                right[i] = n - i
            else:
                right[i] = stack[-1] - i
            
            stack.append(i)
        
        ans = 0
        for i in range(n):
            ans += (left[i] * right[i] * arr[i]) % MOD
        
        return ans